### PR TITLE
Fix bugs handling user data when not logged in

### PR DESCRIPTION
**Pull request by cesmoak**
This fixes two outstanding bugs:
- When the user is not logged in, the list of apps to update is still requested as if there is a logged-in user, breaking when the request comes back with an authentication error
- Notifications can be requested from the server when the user is not logged in. When a user logs in, the notifications may be set to an invalid value until the user's notifications have been returned by the server, breaking the account page

### CODE REVIEW

File: src/app/account/account.component.html [hunk 1]
```diff
@@ -1,130 +1,113 @@
...
-    </div>
-    <div class="waves-effect waves-light btn-flat white-text top-button"
-         mz-tooltip tooltip="Spaces that you have created."
-         position="bottom"
-         routerLink="/account/subscribed-spaces">
-      <i class="material-icons left">location_on</i> Spaces ({{isDev?app_totals.space_total+app_totals.space_sub_total:app_totals.space_sub_total}})
+    <div class="waves-effect waves-light btn-flat white-text top-button" mz-tooltip
+      tooltip="Spaces that you have created." position="bottom" routerLink="/account/friends">
+      <i class="material-icons left">people</i> Sidekicks <span class="new badge new-badge"
+        *ngIf="friendRequestCount() > 0">{{ friendRequestCount() | notOverNine }}</span>
```
Comment (cesmoak) [inline src/app/account/account.component.html]: This is the only non-whitespace change in this file. The use of the function is for cleanliness, but is not necessary for fixing a bug.

File: src/app/account/account.component.ts [hunk 2]
```diff
@@ -1,796 +1,730 @@
...
+        return fetch('https://api.github.com/repos/' + app.github_name + '/' + app.github_repo + '/releases').then(async r => {
+            if (r.ok) {
+                return r.json();
+            }
+        });
+    }
+
+    public friendRequestCount() {
+        return this.appService.notifications.friend_requests.length;
+    }
```
Comment (cesmoak) [inline src/app/account/account.component.ts]: This is the only non-whitespace change in this file.

File: src/app/apps-to-update.service.ts [hunk 3]
```diff
@@ -1,54 +1,52 @@
...
-      }
-    );
-  }
+    constructor(private expanseService: ExpanseClientService, private appService: AppService) {
+        this.expanseService.installedAppsChangedAt.subscribe((_timestamp: number) => {
+            this.updateAppList();
+        });
+        this.appService.authenticationStatus.subscribe((_isAuthenticated: boolean) => {
+            this.updateAppList();
+        });
```
Comment (cesmoak) [inline src/app/apps-to-update.service.ts]: The apps the update list is now also updated when the authentication status changes.

File: src/app/apps-to-update.service.ts [hunk 4]
```diff
@@ -1,54 +1,52 @@
...
-    while (hasMore) {
-      const pageList = await this.getAppsToUpdatePage(page);
-      list.push(...pageList);
-      hasMore = pageList.length > 0;
-      page += 1;
-      return list;
+    private async getAppsToUpdate(): Promise<AppListing[]> {
+        if (!this.appService.isAuthenticated) {
+            return [];
+        }
```
Comment (cesmoak) [inline src/app/apps-to-update.service.ts]: If not authenticated, the request to the server is not performed.

File: src/app/expanse-client.service.ts [hunk 5]
```diff
@@ -1,1035 +1,923 @@
...
+    async getNotifications(): Promise<NotificationsResponse> {
+        const response = (await this.emit('get-notifications', {})) as Record<string, any>;
+        const isError = !response || response.error;
+        if (isError) {
+            throw response;
+        } else {
+            const { friend_requests, unread_messages } = response;
+            return { friend_requests: friend_requests || [], unread_messages: unread_messages || [] };
+        }
+    }
```
Comment (cesmoak) [inline src/app/expanse-client.service.ts]: This method now returns a consistently-shaped object or throws on error.

File: src/app/app.service.ts [hunk 6]
```diff
@@ -1,371 +1,366 @@
...
+    }
+
+    async getNotifications(expanseService: ExpanseClientService) {
+        await expanseService.start();
+        try {
+            this.notifications = await expanseService.getNotifications();
+        } catch {
+            this.notifications = { friend_requests: [], unread_messages: [] };
+        }
+    }
```
Comment (cesmoak) [inline src/app/app.service.ts]: AppService::notifications will now always be set with a consistent shape.

File: src/app/app.service.ts [hunk 7]
```diff
@@ -1,371 +1,366 @@
...
+    sidequestResolve: any;
+    sidequestReject: any;
+    sidequestUrl: string;
+    urlTimeout: any;
+    urlTimeoutValue: number;
+    confirmOpen: MzModalComponent;
+    isGrid: boolean = true;
+    isTimeline: boolean = false;
+    accountComponent: AccountComponent;
+    notifications: Notifications = { friend_requests: [], unread_messages: [] };
```
Comment (cesmoak) [inline src/app/app.service.ts]: Narrowed typing.

File: src/app/app.service.ts [hunk 8]
```diff
@@ -1,371 +1,366 @@
...
+    private _isAuthenticated = false;
+
+    public get isAuthenticated() {
+        return this._isAuthenticated;
+    }
+    public set isAuthenticated(value: boolean) {
+        if (value === this._isAuthenticated) return;
+        this._isAuthenticated = value;
+        this.authenticationStatus.next(this._isAuthenticated);
+    }
```
Comment (cesmoak) [inline src/app/app.service.ts]: An event is now emitted when setting `AppService::isAuthenticated`.

**Comment by shaneharris** [comment]
this looks great, merging in to test locally with a view to pushing as a priority.

**Pull request merged**